lib/elixir/lib/calendar.ex

```diff
@@ -1299,10 +1291,26 @@ defmodule NaiveDateTime do
   def diff(%NaiveDateTime{} = naive_datetime1,
            %NaiveDateTime{} = naive_datetime2,
            unit \\ :second) do
-    ndt1_microsecond = to_microsecond(naive_datetime1)
-    ndt2_microsecond = to_microsecond(naive_datetime2)
-    difference = ndt1_microsecond - ndt2_microsecond
-    System.convert_time_unit(difference, :microsecond, unit)
+    if not Calendar.compatible_calendars?(naive_datetime1.calendar, naive_datetime2.calendar) do
+      raise ArgumentError, "cannot calculate the difference between #{inspect naive_datetime1} and #{inspect naive_datetime2} because their calendars are not compatible and thus the result would be ambiguous"
```

Comment:
Line 605 have the same message, wouldn't it be good to extract to a function? ;)